core/src/main/java/org/infinispan/interceptors/impl/TransactionalStoreInterceptor.java

```diff
@@ -0,0 +1,172 @@
+package org.infinispan.interceptors.impl;
+
+import org.infinispan.atomic.impl.AtomicHashMap;
+import org.infinispan.commands.AbstractVisitor;
+import org.infinispan.commands.tx.CommitCommand;
+import org.infinispan.commands.tx.PrepareCommand;
+import org.infinispan.commands.tx.RollbackCommand;
+import org.infinispan.commands.write.ApplyDeltaCommand;
+import org.infinispan.commands.write.ClearCommand;
+import org.infinispan.commands.write.PutKeyValueCommand;
+import org.infinispan.commands.write.PutMapCommand;
+import org.infinispan.commands.write.RemoveCommand;
+import org.infinispan.commands.write.ReplaceCommand;
+import org.infinispan.commands.write.WriteCommand;
+import org.infinispan.commons.marshall.StreamingMarshaller;
+import org.infinispan.container.InternalEntryFactory;
+import org.infinispan.container.entries.CacheEntry;
+import org.infinispan.container.entries.DeltaAwareCacheEntry;
+import org.infinispan.container.entries.InternalCacheEntry;
+import org.infinispan.container.entries.InternalCacheValue;
+import org.infinispan.context.Flag;
+import org.infinispan.context.InvocationContext;
+import org.infinispan.context.impl.TxInvocationContext;
+import static org.infinispan.factories.KnownComponentNames.CACHE_MARSHALLER;
+import org.infinispan.factories.annotations.ComponentName;
+import org.infinispan.factories.annotations.Inject;
+import org.infinispan.factories.annotations.Start;
+import org.infinispan.interceptors.DDSequentialInterceptor;
+import org.infinispan.marshall.core.MarshalledEntry;
+import org.infinispan.marshall.core.MarshalledEntryImpl;
+import org.infinispan.persistence.PersistenceUtil;
+import static org.infinispan.persistence.PersistenceUtil.internalMetadata;
+import static org.infinispan.persistence.manager.PersistenceManager.AccessMode.BOTH;
+import static org.infinispan.persistence.manager.PersistenceManager.AccessMode.PRIVATE;
+import static org.infinispan.persistence.manager.PersistenceManager.AccessMode.SHARED;
+import org.infinispan.persistence.manager.TransactionalPersistenceManager;
+
+import javax.transaction.Transaction;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * @author Ryan Emerson
+ */
+public class TransactionalStoreInterceptor extends DDSequentialInterceptor {
+   private TransactionalPersistenceManager persistenceManager;
+   private InternalEntryFactory entryFactory;
+   private StreamingMarshaller marshaller;
+   private boolean enabled = true;
+
+   @Inject
+   protected void init(TransactionalPersistenceManager persistenceManager, InternalEntryFactory entryFactory,
+                       @ComponentName(CACHE_MARSHALLER) StreamingMarshaller marshaller) {
+      this.persistenceManager = persistenceManager;
+      this.entryFactory = entryFactory;
+      this.marshaller = marshaller;
+   }
+
+   @Start(priority = 15)
+   protected void start() {
```

Comment:
@wburns actually, you're right :). I was getting ``` java.lang.NoSuchMethodException: java.lang.Object.start()``` before, but that was because of my dev setup.